common/argparse.h

```diff
@@ -0,0 +1,895 @@
+//   - For flags with a default `true` value, they are rendered in help using
+//     the `no-` prefix.
+//
+// - Options may additionally have a short name of a single character [a-zA-Z].
+//   - There is no distinction between the behavior of long and short names.
+//   - The short name can only specify the positive or `true` value for flags.
+//     There is no negative form of short names.
+//   - Short names are parsed after a single `-`, such as `-v`.
+//   - Any number of short names may be grouped after `-`, such as `-xyz`: this
+//     is three options, `x`, `y`, and `z`.
+//   - Short options may include a value after an `=`, but only the last short
+//     option of a group may receive a value. Other short options must have a
+//     default value that will be used. Typically this is used for a set of
+//     flags that are all set to `true`.
+//
+// - Options are parsed from any argument until either a subcommand switches to
+//   that subcommand's options or a `--` argument ends all option parsing to
+//   allow arguments that would be ambiguous with the option syntax.
+//
+// ## Subcommands
+//
+// Subcommands can be nested to any depth, and each subcommand gets its own
+// option space.
+//
+// ## Positional arguments
+//
+// Leaf commands (and subcommands) can accept positional arguments. These work
+// similar to options but consist *only* of the value. They have names, but the
+// name is only used in documentation and error messages. Positional argument
+// values cannot start with a `-` character until after option parsing is ended
+// with a `--` argument.
+//
+// Singular positional arguments store the single value in that position.
+// Appending positional arguments append all of the values in sequence.
+//
+// Multiple positional arguments to a single command are parsed in sequence. For
+// appending positional arguments, the sequence of values appended is ended with
+// a `--` argument. For example, a command that takes two sequences of
+// positional arguments might look like:
+//
+//     my-diff-tool a.txt b.txt c.txt -- new-a.txt new-b.txt new-c.txt
+//
+// The `--` used in this way *both* ends option parsing and the positional
+// argument sequence.
+//
+// ## Help text blocks and rendering
+//
+// At many points in this library, a block of text is specified as a string.
+// These should be written using multi-line raw string literals that start on
+// there own line such as:
+//
+// ```cpp
+//     ...
+//     .help = R"""(
+// Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
+// tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam,
+// quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo
+// consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse
+// cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non
+// proident, sunt in culpa qui officia deserunt mollit anim id est laborum.
```

Comment:
```suggestion
// tempor incididunt ut labore et dolore magna aliqua.
```

Minor: consider giving less boilerplate example text so that more documentation fits on-screen.